scarlet/transformations.py

```diff
@@ -76,27 +76,76 @@ class LinearFilter:
         ----------
-        img: 2D or 1D array-like
-            Weights to apply to the filter
+        values: 2D or 1D array-like
+            Weights to apply to the filter.
+            If `coords` is `None` than either `values` must
+            be 2D with an odd number of rows and columns
+            (and the current pixel in the center) or the
+            location of the current pixel (`center`) in
+            `values` must be specified.
         coords: 2D or 1D array-like
             Relative coordinates from the current pixel
             for each weight in `img`
             (so [0,0] is the current pixel).
+        center: array-like
+            index of the current pixel in `values`.
+            If a set of `coords` is not defined, they can be
+            created as long as either the `center` location in
+            `values` is specified or `values` has an odd number
+            of rows and columns.
+            For example, if `values` is `[[0,1],[2,3]]` and the
+            current pixel is the upper right element, then
+            `center=[0,0]` (`values=0`). If current pixel is the top right then
+            `center=[0,1]` (`values=1`).
         """
-        self.img = np.array(img)
-        self._coords = np.array(coords)
-        self.slices, self.inv_slices = get_filter_slices(self._coords.reshape(-1, 2))
-    
+        self.values = np.array(values)
+        self.update_coords(coords, center)
+
     @property
     def T(self):
         """Transpose the filter
         """
-        return LinearFilter(self.img, -self._coords)
+        return LinearFilter(self.values, -self._coords)
+
+    def update_coords(self, coords=None, center=None):
+        """Update the coordinates for the values
+
+        Create (if necessary) and store the `coords` and `flat_coords`
+        (a flattened version of coords), which map the values to the
+        filter.
+
+        See `__init__` for a description of the parameters
+        """
+        if coords is None:
+            # Attempt to automatically create coordinate grid
+            if len(values.shape)!=2:
+                raise ValueError("Either `values` must be 2D or `coords` must be specified")
+            if center is None:
+                if self.values.shape[0] % 2 == 0 or self.values.shape[1] % 2 == 0:
+                    msg = """Ambiguous center of the `values` array,
+                             you must either specify a set of `coords` or use
+                             a `values` array with an odd number of rows and columns"""
+                    raise ValueError(msg)
+                center = [self.values.shape[0]//2, self.values.shape[1]//2]
+            self.center = center
+            x = np.arange(self.values.shape[1])
+            y = np.arange(self.values.shape[0])
+            x,y = np.meshgrid(x,y)
+            x -= center[1]
+            y -= center[0]
+            coords = np.dstack([y,x])
+        self._coords = np.array(coords)
```

Comment:
Thanks for making the adapter for 2D images.

I would prefer if we only had one version of `values` and `coords` stored in the class, namely the flattened ones. You can regenerate the full ones if needed from them anyway, so no reason to keep them around. I would therefore replace the lines from here to end of method with
```
self._flat_values = np.array(values).reshape(-1)
self._flat_coords = self._coords.reshape(-1,2)
assert ...
# remove elements with zero value
non_zero = self._flat_values != 0
self._flat_values = self._flat_values[non_zero]
self._flat_coords = self._flat_coords[non_zero]
self._slices, self._inv_slices = get_filter_slices(self._flat_coords)
```
and run the entire method as `__init__` because there's nothing left in init to do (since `self.values` isn't needed anymore). Note that the transpose operation could work on the flattened values and coordinates just as well.